extensions-core/caffeine-cache/src/main/java/io/druid/client/cache/CaffeineCache.java

```diff
@@ -0,0 +1,196 @@
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.google.common.primitives.Chars;
+import com.google.common.primitives.Ints;
+import com.metamx.common.logger.Logger;
+import com.metamx.emitter.service.ServiceEmitter;
+import com.metamx.emitter.service.ServiceMetricEvent;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+import net.jpountz.lz4.LZ4Compressor;
+import net.jpountz.lz4.LZ4Factory;
+import net.jpountz.lz4.LZ4FastDecompressor;
+
+public class CaffeineCache implements io.druid.client.cache.Cache
+{
+  private static final Logger log = new Logger(CaffeineCache.class);
+  private static final int FIXED_COST = 8; // Minimum cost in "weight" per entry;
+  private final Cache<NamedKey, byte[]> cache;
+  private final AtomicReference<CacheStats> priorStats = new AtomicReference<>(null);
+  private final CaffeineCacheConfig config;
+
+
+  public static CaffeineCache create(final CaffeineCacheConfig config)
+  {
+    return create(config, config.createExecutor());
+  }
+
+  // Used in testing
+  public static CaffeineCache create(final CaffeineCacheConfig config, @Nullable final Executor executor)
+  {
+    Caffeine<Object, Object> builder = Caffeine.newBuilder().recordStats();
+    if (config.getExpireAfter() >= 0) {
+      builder
+          .expireAfterAccess(config.getExpireAfter(), TimeUnit.MILLISECONDS);
+    }
+    if (config.getSizeInBytes() >= 0) {
+      builder
+          .maximumWeight(config.getSizeInBytes())
+          .weigher((NamedKey key, byte[] value) -> value.length
+                                                   + key.key.length
+                                                   + key.namespace.length() * Chars.BYTES
+                                                   + FIXED_COST);
+    }
+    if (executor != null) {
+      builder.executor(executor);
+    }
+    return new CaffeineCache(builder.build(), config);
+  }
+
+  public CaffeineCache(final Cache<NamedKey, byte[]> cache, CaffeineCacheConfig config)
+  {
+    this.cache = cache;
+    this.config = config;
+  }
+
+  @Override
+  public byte[] get(NamedKey key)
+  {
+    return deserialize(cache.getIfPresent(key));
+  }
+
+  @Override
+  public void put(NamedKey key, byte[] value)
+  {
+    cache.put(key, serialize(value));
+  }
+
+  @Override
+  public Map<NamedKey, byte[]> getBulk(Iterable<NamedKey> keys)
+  {
+    // The assumption here is that every value is accessed at least once. Materializing here ensures deserialize is only
+    // called *once* per value.
+    return ImmutableMap.copyOf(Maps.transformValues(cache.getAllPresent(keys), this::deserialize));
+  }
+
+  // This is completely racy with put. Any values missed should be evicted later anyways. So no worries.
```

Comment:
Is this only here to try to get rid of references sooner rather than later?

Also will later eviction really happen? Is there a background thread doing that?